File: src/frontend/contabil/lib/app/data/model/contabil_parametro_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:contabil/app/data/domain/domain_imports.dart';

class ContabilParametroModel {
	int? id;
	String? mascara;
	int? niveis;
	String? informarContaPor;
	String? compartilhaPlanoConta;
	String? compartilhaHistoricos;
	String? alteraLancamentoOutro;
	String? historicoObrigatorio;
	String? permiteLancamentoZerado;
	String? geraInformativoSped;
	String? spedFormaEscritDiario;
	String? spedNomeLivroDiario;
	String? assinaturaDireita;
	String? assinaturaEsquerda;
	String? contaAtivo;
	String? contaPassivo;
	String? contaPatrimonioLiquido;
	String? contaDepreciacaoAcumulada;
	String? contaCapitalSocial;
	String? contaResultadoExercicio;
	String? contaPrejuizoAcumulado;
	String? contaLucroAcumulado;
	String? contaTituloPagar;
	String? contaTituloReceber;
	String? contaJurosPassivo;
	String? contaJurosAtivo;
	String? contaDescontoObtido;
	String? contaDescontoConcedido;
	String? contaCmv;
	String? contaVenda;
	String? contaVendaServico;
	String? contaEstoque;
	String? contaApuraResultado;
	String? contaJurosApropriar;
	int? idHistPadraoResultado;
	int? idHistPadraoLucro;
	int? idHistPadraoPrejuizo;

	ContabilParametroModel({
		this.id,
		this.mascara,
		this.niveis,
		this.informarContaPor,
		this.compartilhaPlanoConta,
		this.compartilhaHistoricos,
		this.alteraLancamentoOutro,
		this.historicoObrigatorio,
		this.permiteLancamentoZerado,
		this.geraInformativoSped,
		this.spedFormaEscritDiario,
		this.spedNomeLivroDiario,
		this.assinaturaDireita,
		this.assinaturaEsquerda,
		this.contaAtivo,
		this.contaPassivo,
		this.contaPatrimonioLiquido,
		this.contaDepreciacaoAcumulada,
		this.contaCapitalSocial,
		this.contaResultadoExercicio,
		this.contaPrejuizoAcumulado,
		this.contaLucroAcumulado,
		this.contaTituloPagar,
		this.contaTituloReceber,
		this.contaJurosPassivo,
		this.contaJurosAtivo,
		this.contaDescontoObtido,
		this.contaDescontoConcedido,
		this.contaCmv,
		this.contaVenda,
		this.contaVendaServico,
		this.contaEstoque,
		this.contaApuraResultado,
		this.contaJurosApropriar,
		this.idHistPadraoResultado,
		this.idHistPadraoLucro,
		this.idHistPadraoPrejuizo,
	});

	static List<String> dbColumns = <String>[
		'id',
		'mascara',
		'niveis',
		'informar_conta_por',
		'compartilha_plano_conta',
		'compartilha_historicos',
		'altera_lancamento_outro',
		'historico_obrigatorio',
		'permite_lancamento_zerado',
		'gera_informativo_sped',
		'sped_forma_escrit_diario',
		'sped_nome_livro_diario',
		'assinatura_direita',
		'assinatura_esquerda',
		'conta_ativo',
		'conta_passivo',
		'conta_patrimonio_liquido',
		'conta_depreciacao_acumulada',
		'conta_capital_social',
		'conta_resultado_exercicio',
		'conta_prejuizo_acumulado',
		'conta_lucro_acumulado',
		'conta_titulo_pagar',
		'conta_titulo_receber',
		'conta_juros_passivo',
		'conta_juros_ativo',
		'conta_desconto_obtido',
		'conta_desconto_concedido',
		'conta_cmv',
		'conta_venda',
		'conta_venda_servico',
		'conta_estoque',
		'conta_apura_resultado',
		'conta_juros_apropriar',
		'id_hist_padrao_resultado',
		'id_hist_padrao_lucro',
		'id_hist_padrao_prejuizo',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Mascara',
		'Niveis',
		'Informar Conta Por',
		'Compartilha Plano Conta',
		'Compartilha Historicos',
		'Altera Lancamento Outro',
		'Historico Obrigatorio',
		'Permite Lancamento Zerado',
		'Gera Informativo Sped',
		'Sped Forma Escrit Diario',
		'Sped Nome Livro Diario',
		'Assinatura Direita',
		'Assinatura Esquerda',
		'Conta Ativo',
		'Conta Passivo',
		'Conta Patrimonio Liquido',
		'Conta Depreciacao Acumulada',
		'Conta Capital Social',
		'Conta Resultado Exercicio',
		'Conta Prejuizo Acumulado',
		'Conta Lucro Acumulado',
		'Conta Titulo Pagar',
		'Conta Titulo Receber',
		'Conta Juros Passivo',
		'Conta Juros Ativo',
		'Conta Desconto Obtido',
		'Conta Desconto Concedido',
		'Conta Cmv',
		'Conta Venda',
		'Conta Venda Servico',
		'Conta Estoque',
		'Conta Apura Resultado',
		'Conta Juros Apropriar',
		'Id Hist Padrao Resultado',
		'Id Hist Padrao Lucro',
		'Id Hist Padrao Prejuizo',
	];

	ContabilParametroModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		mascara = jsonData['mascara'];
		niveis = jsonData['niveis'];
		informarContaPor = ContabilParametroDomain.getInformarContaPor(jsonData['informarContaPor']);
		compartilhaPlanoConta = ContabilParametroDomain.getCompartilhaPlanoConta(jsonData['compartilhaPlanoConta']);
		compartilhaHistoricos = ContabilParametroDomain.getCompartilhaHistoricos(jsonData['compartilhaHistoricos']);
		alteraLancamentoOutro = ContabilParametroDomain.getAlteraLancamentoOutro(jsonData['alteraLancamentoOutro']);
		historicoObrigatorio = ContabilParametroDomain.getHistoricoObrigatorio(jsonData['historicoObrigatorio']);
		permiteLancamentoZerado = ContabilParametroDomain.getPermiteLancamentoZerado(jsonData['permiteLancamentoZerado']);
		geraInformativoSped = ContabilParametroDomain.getGeraInformativoSped(jsonData['geraInformativoSped']);
		spedFormaEscritDiario = ContabilParametroDomain.getSpedFormaEscritDiario(jsonData['spedFormaEscritDiario']);
		spedNomeLivroDiario = jsonData['spedNomeLivroDiario'];
		assinaturaDireita = jsonData['assinaturaDireita'];
		assinaturaEsquerda = jsonData['assinaturaEsquerda'];
		contaAtivo = jsonData['contaAtivo'];
		contaPassivo = jsonData['contaPassivo'];
		contaPatrimonioLiquido = jsonData['contaPatrimonioLiquido'];
		contaDepreciacaoAcumulada = jsonData['contaDepreciacaoAcumulada'];
		contaCapitalSocial = jsonData['contaCapitalSocial'];
		contaResultadoExercicio = jsonData['contaResultadoExercicio'];
		contaPrejuizoAcumulado = jsonData['contaPrejuizoAcumulado'];
		contaLucroAcumulado = jsonData['contaLucroAcumulado'];
		contaTituloPagar = jsonData['contaTituloPagar'];
		contaTituloReceber = jsonData['contaTituloReceber'];
		contaJurosPassivo = jsonData['contaJurosPassivo'];
		contaJurosAtivo = jsonData['contaJurosAtivo'];
		contaDescontoObtido = jsonData['contaDescontoObtido'];
		contaDescontoConcedido = jsonData['contaDescontoConcedido'];
		contaCmv = jsonData['contaCmv'];
		contaVenda = jsonData['contaVenda'];
		contaVendaServico = jsonData['contaVendaServico'];
		contaEstoque = jsonData['contaEstoque'];
		contaApuraResultado = jsonData['contaApuraResultado'];
		contaJurosApropriar = jsonData['contaJurosApropriar'];
		idHistPadraoResultado = jsonData['idHistPadraoResultado'];
		idHistPadraoLucro = jsonData['idHistPadraoLucro'];
		idHistPadraoPrejuizo = jsonData['idHistPadraoPrejuizo'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['mascara'] = mascara;
		jsonData['niveis'] = niveis;
		jsonData['informarContaPor'] = ContabilParametroDomain.setInformarContaPor(informarContaPor);
		jsonData['compartilhaPlanoConta'] = ContabilParametroDomain.setCompartilhaPlanoConta(compartilhaPlanoConta);
		jsonData['compartilhaHistoricos'] = ContabilParametroDomain.setCompartilhaHistoricos(compartilhaHistoricos);
		jsonData['alteraLancamentoOutro'] = ContabilParametroDomain.setAlteraLancamentoOutro(alteraLancamentoOutro);
		jsonData['historicoObrigatorio'] = ContabilParametroDomain.setHistoricoObrigatorio(historicoObrigatorio);
		jsonData['permiteLancamentoZerado'] = ContabilParametroDomain.setPermiteLancamentoZerado(permiteLancamentoZerado);
		jsonData['geraInformativoSped'] = ContabilParametroDomain.setGeraInformativoSped(geraInformativoSped);
		jsonData['spedFormaEscritDiario'] = ContabilParametroDomain.setSpedFormaEscritDiario(spedFormaEscritDiario);
		jsonData['spedNomeLivroDiario'] = spedNomeLivroDiario;
		jsonData['assinaturaDireita'] = assinaturaDireita;
		jsonData['assinaturaEsquerda'] = assinaturaEsquerda;
		jsonData['contaAtivo'] = contaAtivo;
		jsonData['contaPassivo'] = contaPassivo;
		jsonData['contaPatrimonioLiquido'] = contaPatrimonioLiquido;
		jsonData['contaDepreciacaoAcumulada'] = contaDepreciacaoAcumulada;
		jsonData['contaCapitalSocial'] = contaCapitalSocial;
		jsonData['contaResultadoExercicio'] = contaResultadoExercicio;
		jsonData['contaPrejuizoAcumulado'] = contaPrejuizoAcumulado;
		jsonData['contaLucroAcumulado'] = contaLucroAcumulado;
		jsonData['contaTituloPagar'] = contaTituloPagar;
		jsonData['contaTituloReceber'] = contaTituloReceber;
		jsonData['contaJurosPassivo'] = contaJurosPassivo;
		jsonData['contaJurosAtivo'] = contaJurosAtivo;
		jsonData['contaDescontoObtido'] = contaDescontoObtido;
		jsonData['contaDescontoConcedido'] = contaDescontoConcedido;
		jsonData['contaCmv'] = contaCmv;
		jsonData['contaVenda'] = contaVenda;
		jsonData['contaVendaServico'] = contaVendaServico;
		jsonData['contaEstoque'] = contaEstoque;
		jsonData['contaApuraResultado'] = contaApuraResultado;
		jsonData['contaJurosApropriar'] = contaJurosApropriar;
		jsonData['idHistPadraoResultado'] = idHistPadraoResultado;
		jsonData['idHistPadraoLucro'] = idHistPadraoLucro;
		jsonData['idHistPadraoPrejuizo'] = idHistPadraoPrejuizo;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		mascara = plutoRow.cells['mascara']?.value;
		niveis = plutoRow.cells['niveis']?.value;
		informarContaPor = plutoRow.cells['informarContaPor']?.value != '' ? plutoRow.cells['informarContaPor']?.value : 'Código';
		compartilhaPlanoConta = plutoRow.cells['compartilhaPlanoConta']?.value != '' ? plutoRow.cells['compartilhaPlanoConta']?.value : 'Sim';
		compartilhaHistoricos = plutoRow.cells['compartilhaHistoricos']?.value != '' ? plutoRow.cells['compartilhaHistoricos']?.value : 'Sim';
		alteraLancamentoOutro = plutoRow.cells['alteraLancamentoOutro']?.value != '' ? plutoRow.cells['alteraLancamentoOutro']?.value : 'Sim';
		historicoObrigatorio = plutoRow.cells['historicoObrigatorio']?.value != '' ? plutoRow.cells['historicoObrigatorio']?.value : 'Sim';
		permiteLancamentoZerado = plutoRow.cells['permiteLancamentoZerado']?.value != '' ? plutoRow.cells['permiteLancamentoZerado']?.value : 'Sim';
		geraInformativoSped = plutoRow.cells['geraInformativoSped']?.value != '' ? plutoRow.cells['geraInformativoSped']?.value : 'Sim';
		spedFormaEscritDiario = plutoRow.cells['spedFormaEscritDiario']?.value != '' ? plutoRow.cells['spedFormaEscritDiario']?.value : 'Livro Diário Completo';
		spedNomeLivroDiario = plutoRow.cells['spedNomeLivroDiario']?.value;
		assinaturaDireita = plutoRow.cells['assinaturaDireita']?.value;
		assinaturaEsquerda = plutoRow.cells['assinaturaEsquerda']?.value;
		contaAtivo = plutoRow.cells['contaAtivo']?.value;
		contaPassivo = plutoRow.cells['contaPassivo']?.value;
		contaPatrimonioLiquido = plutoRow.cells['contaPatrimonioLiquido']?.value;
		contaDepreciacaoAcumulada = plutoRow.cells['contaDepreciacaoAcumulada']?.value;
		contaCapitalSocial = plutoRow.cells['contaCapitalSocial']?.value;
		contaResultadoExercicio = plutoRow.cells['contaResultadoExercicio']?.value;
		contaPrejuizoAcumulado = plutoRow.cells['contaPrejuizoAcumulado']?.value;
		contaLucroAcumulado = plutoRow.cells['contaLucroAcumulado']?.value;
		contaTituloPagar = plutoRow.cells['contaTituloPagar']?.value;
		contaTituloReceber = plutoRow.cells['contaTituloReceber']?.value;
		contaJurosPassivo = plutoRow.cells['contaJurosPassivo']?.value;
		contaJurosAtivo = plutoRow.cells['contaJurosAtivo']?.value;
		contaDescontoObtido = plutoRow.cells['contaDescontoObtido']?.value;
		contaDescontoConcedido = plutoRow.cells['contaDescontoConcedido']?.value;
		contaCmv = plutoRow.cells['contaCmv']?.value;
		contaVenda = plutoRow.cells['contaVenda']?.value;
		contaVendaServico = plutoRow.cells['contaVendaServico']?.value;
		contaEstoque = plutoRow.cells['contaEstoque']?.value;
		contaApuraResultado = plutoRow.cells['contaApuraResultado']?.value;
		contaJurosApropriar = plutoRow.cells['contaJurosApropriar']?.value;
		idHistPadraoResultado = plutoRow.cells['idHistPadraoResultado']?.value;
		idHistPadraoLucro = plutoRow.cells['idHistPadraoLucro']?.value;
		idHistPadraoPrejuizo = plutoRow.cells['idHistPadraoPrejuizo']?.value;
	}	

	ContabilParametroModel clone() {
		return ContabilParametroModel(
			id: id,
			mascara: mascara,
			niveis: niveis,
			informarContaPor: informarContaPor,
			compartilhaPlanoConta: compartilhaPlanoConta,
			compartilhaHistoricos: compartilhaHistoricos,
			alteraLancamentoOutro: alteraLancamentoOutro,
			historicoObrigatorio: historicoObrigatorio,
			permiteLancamentoZerado: permiteLancamentoZerado,
			geraInformativoSped: geraInformativoSped,
			spedFormaEscritDiario: spedFormaEscritDiario,
			spedNomeLivroDiario: spedNomeLivroDiario,
			assinaturaDireita: assinaturaDireita,
			assinaturaEsquerda: assinaturaEsquerda,
			contaAtivo: contaAtivo,
			contaPassivo: contaPassivo,
			contaPatrimonioLiquido: contaPatrimonioLiquido,
			contaDepreciacaoAcumulada: contaDepreciacaoAcumulada,
			contaCapitalSocial: contaCapitalSocial,
			contaResultadoExercicio: contaResultadoExercicio,
			contaPrejuizoAcumulado: contaPrejuizoAcumulado,
			contaLucroAcumulado: contaLucroAcumulado,
			contaTituloPagar: contaTituloPagar,
			contaTituloReceber: contaTituloReceber,
			contaJurosPassivo: contaJurosPassivo,
			contaJurosAtivo: contaJurosAtivo,
			contaDescontoObtido: contaDescontoObtido,
			contaDescontoConcedido: contaDescontoConcedido,
			contaCmv: contaCmv,
			contaVenda: contaVenda,
			contaVendaServico: contaVendaServico,
			contaEstoque: contaEstoque,
			contaApuraResultado: contaApuraResultado,
			contaJurosApropriar: contaJurosApropriar,
			idHistPadraoResultado: idHistPadraoResultado,
			idHistPadraoLucro: idHistPadraoLucro,
			idHistPadraoPrejuizo: idHistPadraoPrejuizo,
		);			
	}

	
}

File: backend/contabil/src/model/contabil_parametro_model.py
from src import db


class ContabilParametroModel(db.Model):
    __tablename__ = 'contabil_parametro'

    id = db.Column(db.Integer, primary_key=True)
    mascara = db.Column(db.String(30))
    niveis = db.Column(db.Integer)
    informar_conta_por = db.Column(db.String(1))
    compartilha_plano_conta = db.Column(db.String(1))
    compartilha_historicos = db.Column(db.String(1))
    altera_lancamento_outro = db.Column(db.String(1))
    historico_obrigatorio = db.Column(db.String(1))
    permite_lancamento_zerado = db.Column(db.String(1))
    gera_informativo_sped = db.Column(db.String(1))
    sped_forma_escrit_diario = db.Column(db.String(1))
    sped_nome_livro_diario = db.Column(db.String(100))
    assinatura_direita = db.Column(db.Text)
    assinatura_esquerda = db.Column(db.Text)
    conta_ativo = db.Column(db.String(30))
    conta_passivo = db.Column(db.String(30))
    conta_patrimonio_liquido = db.Column(db.String(30))
    conta_depreciacao_acumulada = db.Column(db.String(30))
    conta_capital_social = db.Column(db.String(30))
    conta_resultado_exercicio = db.Column(db.String(30))
    conta_prejuizo_acumulado = db.Column(db.String(30))
    conta_lucro_acumulado = db.Column(db.String(30))
    conta_titulo_pagar = db.Column(db.String(30))
    conta_titulo_receber = db.Column(db.String(30))
    conta_juros_passivo = db.Column(db.String(30))
    conta_juros_ativo = db.Column(db.String(30))
    conta_desconto_obtido = db.Column(db.String(30))
    conta_desconto_concedido = db.Column(db.String(30))
    conta_cmv = db.Column(db.String(30))
    conta_venda = db.Column(db.String(30))
    conta_venda_servico = db.Column(db.String(30))
    conta_estoque = db.Column(db.String(30))
    conta_apura_resultado = db.Column(db.String(30))
    conta_juros_apropriar = db.Column(db.String(30))
    id_hist_padrao_resultado = db.Column(db.Integer)
    id_hist_padrao_lucro = db.Column(db.Integer)
    id_hist_padrao_prejuizo = db.Column(db.Integer)


    def mapping(self, data):
        self.id = data.get('id')
        self.mascara = data.get('mascara')
        self.niveis = data.get('niveis')
        self.informar_conta_por = data.get('informarContaPor')
        self.compartilha_plano_conta = data.get('compartilhaPlanoConta')
        self.compartilha_historicos = data.get('compartilhaHistoricos')
        self.altera_lancamento_outro = data.get('alteraLancamentoOutro')
        self.historico_obrigatorio = data.get('historicoObrigatorio')
        self.permite_lancamento_zerado = data.get('permiteLancamentoZerado')
        self.gera_informativo_sped = data.get('geraInformativoSped')
        self.sped_forma_escrit_diario = data.get('spedFormaEscritDiario')
        self.sped_nome_livro_diario = data.get('spedNomeLivroDiario')
        self.assinatura_direita = data.get('assinaturaDireita')
        self.assinatura_esquerda = data.get('assinaturaEsquerda')
        self.conta_ativo = data.get('contaAtivo')
        self.conta_passivo = data.get('contaPassivo')
        self.conta_patrimonio_liquido = data.get('contaPatrimonioLiquido')
        self.conta_depreciacao_acumulada = data.get('contaDepreciacaoAcumulada')
        self.conta_capital_social = data.get('contaCapitalSocial')
        self.conta_resultado_exercicio = data.get('contaResultadoExercicio')
        self.conta_prejuizo_acumulado = data.get('contaPrejuizoAcumulado')
        self.conta_lucro_acumulado = data.get('contaLucroAcumulado')
        self.conta_titulo_pagar = data.get('contaTituloPagar')
        self.conta_titulo_receber = data.get('contaTituloReceber')
        self.conta_juros_passivo = data.get('contaJurosPassivo')
        self.conta_juros_ativo = data.get('contaJurosAtivo')
        self.conta_desconto_obtido = data.get('contaDescontoObtido')
        self.conta_desconto_concedido = data.get('contaDescontoConcedido')
        self.conta_cmv = data.get('contaCmv')
        self.conta_venda = data.get('contaVenda')
        self.conta_venda_servico = data.get('contaVendaServico')
        self.conta_estoque = data.get('contaEstoque')
        self.conta_apura_resultado = data.get('contaApuraResultado')
        self.conta_juros_apropriar = data.get('contaJurosApropriar')
        self.id_hist_padrao_resultado = data.get('idHistPadraoResultado')
        self.id_hist_padrao_lucro = data.get('idHistPadraoLucro')
        self.id_hist_padrao_prejuizo = data.get('idHistPadraoPrejuizo')

    def serialize(self):
        return {
            'id': self.id,
            'mascara': self.mascara,
            'niveis': self.niveis,
            'informarContaPor': self.informar_conta_por,
            'compartilhaPlanoConta': self.compartilha_plano_conta,
            'compartilhaHistoricos': self.compartilha_historicos,
            'alteraLancamentoOutro': self.altera_lancamento_outro,
            'historicoObrigatorio': self.historico_obrigatorio,
            'permiteLancamentoZerado': self.permite_lancamento_zerado,
            'geraInformativoSped': self.gera_informativo_sped,
            'spedFormaEscritDiario': self.sped_forma_escrit_diario,
            'spedNomeLivroDiario': self.sped_nome_livro_diario,
            'assinaturaDireita': self.assinatura_direita,
            'assinaturaEsquerda': self.assinatura_esquerda,
            'contaAtivo': self.conta_ativo,
            'contaPassivo': self.conta_passivo,
            'contaPatrimonioLiquido': self.conta_patrimonio_liquido,
            'contaDepreciacaoAcumulada': self.conta_depreciacao_acumulada,
            'contaCapitalSocial': self.conta_capital_social,
            'contaResultadoExercicio': self.conta_resultado_exercicio,
            'contaPrejuizoAcumulado': self.conta_prejuizo_acumulado,
            'contaLucroAcumulado': self.conta_lucro_acumulado,
            'contaTituloPagar': self.conta_titulo_pagar,
            'contaTituloReceber': self.conta_titulo_receber,
            'contaJurosPassivo': self.conta_juros_passivo,
            'contaJurosAtivo': self.conta_juros_ativo,
            'contaDescontoObtido': self.conta_desconto_obtido,
            'contaDescontoConcedido': self.conta_desconto_concedido,
            'contaCmv': self.conta_cmv,
            'contaVenda': self.conta_venda,
            'contaVendaServico': self.conta_venda_servico,
            'contaEstoque': self.conta_estoque,
            'contaApuraResultado': self.conta_apura_resultado,
            'contaJurosApropriar': self.conta_juros_apropriar,
            'idHistPadraoResultado': self.id_hist_padrao_resultado,
            'idHistPadraoLucro': self.id_hist_padrao_lucro,
            'idHistPadraoPrejuizo': self.id_hist_padrao_prejuizo,
        }